Add travis-ci config file

andreas-bok-sociomantic opened the pull request:
Currently test db setup depends on ENVIRON VARS

Running tests, thus requires #7 to be resolved

Addresses #6

andreas-bok-sociomantic commented:
Adding travis config is a bit of a trial and error process. 
Installing dependencies on in the container running the tests require installing a package, `libcloog-ppl0`, which is currently not on the travis whitelist. 
Created an issue for adding this at https://github.com/travis-ci/apt-package-whitelist/issues/2134

andreas-bok-sociomantic commented:
The configuration is atm targeted at 
`ubuntu 14.04 trusty`. 

Issues were related to testing on travis precise: 
1. `automake` was `v1.11`, doesn't support `[serial-test]` as 
test-harness
2. An compile error which was AFAICS caused by the provided GCC v4.6.2 having been installed without support for Graphite loop optimizations. 
> libdrizzle/query.cc:149:6: sorry, unimplemented: Graphite loop optimizations can only be used if the libcloog-ppl0 package is installed

andreas-bok-sociomantic commented:
The project compiles and passes tests with the configuration defined in `.travis.yml`. 

`binlog` tests are ignored since it requires `mysql` to be started with `--log-bin=path-to-binlog-file`

ben-palmer-sociomantic commented:
LGTM, maybe squash the commits and then I can rebase

andreas-bok-sociomantic commented:
I just wanna see if I can set up mysql with binlog enabled before rebase

ben-palmer-sociomantic commented:
> I just wanna see if I can set up mysql with binlog enabled before rebase

Ok, no problem. Let me know when you want me to have another look at it.

andreas-bok-sociomantic commented:
The travis configuration now compiles and passes all tests 
Rebased on upstream/master.
Ready for review

ben-palmer-sociomantic commented:
I am happy with this method of doing the tests. @leandro-lucarella-sociomantic any comments?

leandro-lucarella-sociomantic commented:
Yes: Niiiice!

andreas-bok-sociomantic commented:
@leandro-lucarella-sociomantic do you have another suggestion for changing the mysql config. The `sed` inplace substitute is not the most elegant/fail-safe solution I reckon...

leandro-lucarella-sociomantic commented:
Not really. Is not ideal but I think it's good enough. Will you find out if the file changes, and the `sed` command does nothing? If that's the case, then is a safe hack. If not, we might want to add some check to make sure we are either really testing this, or the tests get totally broken. What we need to avoid is silent breaking and suddenly thinking all is good and nothing is being properly tested.

andreas-bok-sociomantic commented:
> Will you find out if the file changes, and the sed command does nothing?

Made a `diff` against a copy before `sed` of `my.cnf`. So the `sed` does change the mysql config 
```sh
$ sudo service mysql stop
mysql stop/waiting
before_script.2
$ sudo cp /etc/mysql/my.cnf /etc/mysql/my.bck
before_script.3
$ sudo sed  -i 's/\[mysqld\]/[mysqld]\nserver-id = 1\nlog-bin = mysqlbin\nbinlog-format = ROW\n/g' /etc/mysql/my.cnf
$ diff /etc/mysql/my.cnf /etc/mysql/my.bck
32,35d31
< server-id = 1
< log-bin = mysqlbin
< binlog-format = ROW
```

>  What we need to avoid is silent breaking and suddenly thinking all is good and nothing is being properly tested.

Tests fail silently if the fail is due to missing feature support in build environment. 
@leandro-lucarella-sociomantic Given this behaviour, would you recommend breaking the tests on first error by changing the source of the test cases?

```
tests/unit/binlog.cc:81: main SKIP 'ret == DRIZZLE_RETURN_ERROR_CODE' [ Binlog is not open?: Binary log is not open(DRIZZLE_RETURN_ERROR_CODE) ]
SKIP: tests/unit/binlog
PASS: tests/unit/escape
PASS: tests/unit/insert_id
PASS: tests/unit/query
PASS: tests/unit/column
...
tests/unit/datetypes.c:90: main SKIP 'drizzle_server_version_number(con) < 50604' [ Test requires MySQL 5.6.4 or higher ]
SKIP: tests/unit/datetypes
...
======================
All 14 tests passed
(2 tests were not run)
```

ben-palmer-sociomantic commented:
>  Given this behaviour, would you recommend breaking the tests on first error by changing the source of the test cases?

That seems like a good idea to me. If the tests need to be skipped then they should fail.

andreas-bok-sociomantic commented:
Whether tests should fail on first error or silently can now be set via env var 
`SKIP_TEST_ON_ERROR`
Setting the var is optional as it evaluates to false if it is defined.

ben-palmer-sociomantic commented:
> Setting the var is optional as it evaluates to false if it is defined.

Should this be "if it is NOT defined"?

andreas-bok-sociomantic commented:
Yes, typo.

ben-palmer-sociomantic commented:
All the whitespace changes should be in their own commit then LGTM

andreas-bok-sociomantic committed:
Set up travis continuous integration testing

The configuration is targeted at Ubuntu 14.04 (Trusty).
This is due to issues were related to building on the
Travis 12.04 (Precise) build environment:

1. automake was v1.11, doesn't support [serial-test]
   as test-harness
2. A compile error which was AFAICS caused by the
   provided GCC v4.6.2 having been installed without
   support for Graphite loop optimizations.

Environment variables for MYSQL are necessary for
running database test.

```
\#required
MYSQL_SERVER=127.0.0.1
MYSQL_USER=root
MYSQL_PASSWORD=
```

Furthermore an environment variable can be set which
tests breaks on first error or fails silently

```
\#optional
\#if NOT defined it evaluates to 0
SKIP_TEST_ON_ERROR=0 [0|1]
```

Addresses #6

Set environment variables in .travis.yml

andreas-bok-sociomantic commented:
Removed whitespace changes caused by Sublime's, otherwise, helpful trailing whitespace trimming

ben-palmer-sociomantic commented:
ok, LGTM

andreas-bok-sociomantic commented:
@leandro-lucarella-sociomantic any comments?

leandro-lucarella-sociomantic commented:
> Made a diff against a copy before sed of my.cnf. So the sed does change the mysql config

Sorry, what I meant was what would happen if the config file in travis changes and suddenly the sed pattern stop working so it does nothing. Would we detect that problem via failing tests?

leandro-lucarella-sociomantic commented:
> Tests fail silently if the error is due to missing feature support in build environment.
> @leandro-lucarella-sociomantic Given this behaviour, would you recommend breaking the tests on first error by changing the source of the test cases?

Yes, I think that's a good idea. It might be more annoying for users wanting to test, since they'll need to have a complete environment to test, but I guess is a fair thing to ask, and worse case you can just use travis for a complete test.

leandro-lucarella-sociomantic commented on .travis.yml:
```diff
@@ -0,0 +1,54 @@
+  - mysql
+
+## The following environment variables are specified in 
+#  the repository settings at travis-ci.org
```
Any reason why not to specify them here, so any developer can modify them via a PR?

leandro-lucarella-sociomantic commented on yatl/lite.h:
```diff
@@ -52,8 +52,14 @@
+
 #ifndef EXIT_SKIP
-# define EXIT_SKIP 77
+ #define EXIT_SKIP SKIP_TEST_ON_ERROR ? 77 : EXIT_FAILURE
```
Wait, if some tests are skipped then the test program exits with a code 77 and not 0? If so the tests should fail already!

andreas-bok-sociomantic commented:
> Sorry, what I meant was what would happen if the config file in travis changes and suddenly the sed pattern stop working so it does nothing. Would we detect that problem via failing tests?

We would get an error iff `SKIP_TEST_ON_ERROR=1`. The binlog test depends on MySQL having replication enabled. As of now the `sed` substitution  in `my.cnf` is doing that.
However we should accommodate the script ifthe format of the mysql config file changes in the future.

andreas-bok-sociomantic commented:
> Any reason why not to specify them here, so any developer can modify them via a PR?

Admittedly I did exactly to avoid doing a PR when testing the env variables. However defining the variables in the `.travis.yml` file overrules the values defined in the settings at `travis-ci.org`
So I will uncomment the vars in the config file.

andreas-bok-sociomantic commented:
> Wait, if some tests are skipped then the test program exits with a code 77 and not 0? If so the tests 
should fail already!

Yes. but the unittests are separate programs so the unittest report AFAICS simply gathers the return code of the individual tests. 

However with `SKIP_TEST_ON_ERROR=0` skipped tests return `EXIT_FAILURE`. So the travis build will show build as failed in this case

bokchan commented:
test ping @andreas-bok-sociomantic

bokchan commented:
@andreas-bok-sociomantic

andreas-bok-sociomantic commented:
Added environment variables to `.travis.yml` file

leandro-lucarella-sociomantic commented:
> Yes. but the unittests are separate programs so the unittest report AFAICS simply gathers the return code of the individual tests.
> 
> However with SKIP_TEST_ON_ERROR=0 skipped tests return EXIT_FAILURE. So the travis build will show build as failed in this case

Let me see if I understand this correctly, so if the program returns 77, so test runner picks up that return code as skipped and won't fail the overall test?

andreas-bok-sociomantic commented:
> Let me see if I understand this correctly, so if the program returns 77, so test runner picks up that return code as skipped and won't fail the overall test?

As such we could hard code `EXIT_SKIP` as 1 and any skipped test would cause the overall test to fail. However I wanted to make it possible to decide skipped tests should cause the overall test should fail. Hence the env var `SKIP_TEST_ON_ERROR`

andreas-bok-sociomantic commented:
Squashed commits. Should be ready for merge if there's no further comments

ben-palmer-sociomantic commented:
This pull request has been rebased via `git hub pull rebase`. Original pull request HEAD was 6bfd9c6239e4642599f5424c4c70d303fdfb0a70, new (rebased) HEAD is 4df8e0679a93e3ffc2959cef636395d8126e04f1
